examples/rdkafka_idempotent_producer.c

```diff
@@ -0,0 +1,342 @@
+         *       this call.
+         */
+        rk = rd_kafka_new(RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr));
+        if (!rk) {
+                fprintf(stderr,
+                        "%% Failed to create new producer: %s\n", errstr);
+                return 1;
+        }
+
+        /* Signal handler for clean shutdown */
+        signal(SIGINT, stop);
+
+        fprintf(stderr, "%% Running producer loop. Press Ctrl-C to exit\n");
+
+        while (run) {
+                char buf[64];
+
+                snprintf(buf, sizeof(buf),
+                         "Idempotent Producer example message #%d", msgcnt);
+
+                /*
+                 * Produce message.
+                 * This is an asynchronous call, on success it will only
+                 * enqueue the message on the internal producer queue.
+                 * The actual delivery attempts to the broker are handled
+                 * by background threads.
+                 * The previously registered delivery report callback
+                 * (dr_msg_cb) is used to signal back to the application
+                 * when the message has been delivered (or failed),
+                 * and is triggered when the application calls
+                 * rd_kafka_poll() or rd_kafka_flush().
+                 */
+        retry:
+                err = rd_kafka_producev(
+                        rk,
+                        RD_KAFKA_V_TOPIC(topic),
+                        RD_KAFKA_V_VALUE(buf, strlen(buf)),
+                        /* Copy the message payload so the `buf` can
+                         * be reused for the next message. */
+                        RD_KAFKA_V_MSGFLAGS(RD_KAFKA_MSG_F_COPY),
+                        RD_KAFKA_V_END);
+
+                if (err) {
+                        /**
+                         * Failed to *enqueue* message for producing.
+                         */
+                        fprintf(stderr,
+                                "%% Failed to produce to topic %s: %s\n",
+                                topic, rd_kafka_err2str(err));
+
+                        if (err == RD_KAFKA_RESP_ERR__QUEUE_FULL) {
+                                /* If the internal queue is full, wait for
+                                 * messages to be delivered and then retry.
+                                 * The internal queue represents both
+                                 * messages to be sent and messages that have
+                                 * been sent or failed, awaiting their
+                                 * delivery report callback to be called.
+                                 *
+                                 * The internal queue is limited by the
+                                 * configuration property
+                                 * queue.buffering.max.messages */
+                                rd_kafka_poll(rk, 1000/*block for max 1000ms*/);
+                                goto retry;
```

Comment:
Interesting that you prefer goto. seems justifiable - one level less nesting, less loc, not spaghetti here.

Reply:
Yep, those are the reasons. :+1: